server/lib/scheduler/Scheduler.ts

```diff
@@ -0,0 +1,110 @@
+    const assignedWorkspaceId = this.pickWorkspaceIdAtRandom(idsOfWorkspacesThatCouldBeNext);
+    await this.schedule.assignWorkspaceToUser(userId, assignedWorkspaceId);
+  }
+
+  private async getIdsOfWorkspacesThatCouldBeNext(userId) {
+    const allWorkspaces = await this.fetchAllWorkspaces();
+    const allEligibleWorkspaces = await this.filterByEligibility(allWorkspaces);
+    const workspacesInTreeWorkedOnLeastRecently = await this.filterByWhetherInTreeWorkedOnLeastRecently(allEligibleWorkspaces);
+    const notYetWorkedOnInThatTree = await this.filterByWhetherNotYetWorkedOn(workspacesInTreeWorkedOnLeastRecently);
+
+    let finalWorkspaces = notYetWorkedOnInThatTree;
+
+    // if every workspace in that tree has been worked on
+    // then instead look for workspaces with remaining budgets
+    if (finalWorkspaces.length === 0) {
+      finalWorkspaces = await this.filterByWhetherHasRemainingBudget(workspacesInTreeWorkedOnLeastRecently);
+    }
+
+    // if every workspace in that tree has 0 remaining budget
+    // then instead look for the workspace worked on least recently
+    if (finalWorkspaces.length === 0) {
+      const idOfWorkspaceWorkedOnLeastRecently = await this.schedule.getLeastRecentlyActiveWorkspace(workspacesInTreeWorkedOnLeastRecently.map(w => w.id));
+      return [idOfWorkspaceWorkedOnLeastRecently];
+    }
+
+    return finalWorkspaces.map(w => w.id);
+  }
+
+  private pickWorkspaceIdAtRandom(workspaceIds) {
+    const randomIndex = Math.floor(Math.random() * (workspaceIds.length));
+    return workspaceIds[randomIndex];
+  }
+
+  private async filterByEligibility(workspaces) {
+    // use for... of instead of asyncro's filter
+    // because isWorkspaceEligible will use rootParentCache
+    // and the parallel nature of filter doesn't work with
+    // rootParentCache's caching
+    let allEligibleWorkspaces = [];
+    for (const workspace of workspaces) {
+      const isMarkedEligible = await this.isWorkspaceEligible(workspace);
+      const isCurrentlyBeingWorkedOn = await this.schedule.isWorkspaceCurrentlyBeingWorkedOn(workspace.id);
+      if (isMarkedEligible && !isCurrentlyBeingWorkedOn) {
+        allEligibleWorkspaces.push(workspace);
+      }
+    }
+    return allEligibleWorkspaces;
+  }
+
+  /*
+    NOTE: this only considers the trees that at least one workspace in the
+    workspaces argument belong to
+  */
+  private async filterByWhetherInTreeWorkedOnLeastRecently(workspaces) {
+    let workspacesInTreeWorkedOnLeastRecently = [];
+
+    for (const workspace of workspaces) {
+      const shouldInclude = await this.schedule.isInTreeWorkedOnLeastRecently(workspaces.map(w => w.id), workspace.id);
+      if (shouldInclude) {
+        workspacesInTreeWorkedOnLeastRecently.push(workspace);
+      }
+    }
+
+    return workspacesInTreeWorkedOnLeastRecently;
+  }
+
+  private async filterByWhetherNotYetWorkedOn(workspaces) {
```

Comment:
After some thinking, I think one decent way to do this filtering would be with Ramda.js. You should be able to compose multiple filtering methods. 

One thing is that instead of having these filter methods that accept arrays of workspaces, I suggest having filter methods that work on individual workspaces. ``fooWorkspaces.filter(WhetherNotYetWorkedOn)`` and similar seem good to me, and with Ramda would be composable (so you could add multiple filters). Let me know if you would like help with that.